embed/ephy-fx-password-notification.c

```diff
@@ -0,0 +1,162 @@
+/* -*- Mode: C; tab-width: 2; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
+/*
+ *  Copyright © 2016 Gabriel Ivascu <ivascu.gabriel59@gmail.com>
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2, or (at your option)
+ *  any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "config.h"
+#include "ephy-fx-password-notification.h"
+
+#include "ephy-notification-manager.h"
+
+#include <glib/gi18n.h>
+
+struct _EphyFxPasswordNotification {
+  GtkGrid                  parent_instance;
+
+  GtkWidget               *note;
+  GtkWidget               *suggestion;
+
+  const char              *user;
+};
+
+struct _EphyFxPasswordNotificationClass {
+  GtkGridClass parent_class;
+};
+
+enum {
+  PROP_0,
+  PROP_USER
+};
+
+G_DEFINE_TYPE (EphyFxPasswordNotification, ephy_fx_password_notification, GTK_TYPE_GRID);
+
+static void
+ephy_fx_password_notification_constructed (GObject *object)
+{
+  EphyFxPasswordNotification *self = EPHY_FX_PASSWORD_NOTIFICATION (object);
+  char *text;
+
+  text = g_strdup_printf (_("We noticed that the password of your Firefox "
+                            "Account <b>%s</b> has been changed."),
+                          self->user);
+  gtk_label_set_markup (GTK_LABEL (self->note), text);
+  gtk_label_set_text (GTK_LABEL (self->suggestion),
+                      _("Please visit Preferences and sign in with the new "
+                        "password to continue the sync process."));
+  g_free (text);
+
+  G_OBJECT_CLASS (ephy_fx_password_notification_parent_class)->constructed (object);
+}
+
+static void
+ephy_fx_password_notification_dispose (GObject *object)
+{
+  EphyFxPasswordNotification *self = EPHY_FX_PASSWORD_NOTIFICATION (object);
+
+  g_clear_pointer (&self->user, g_free);
```

Comment:
This doesn't need to be done in dispose; dispose exists to break reference cycles, so it's usually where you want to unref objects (with `g_clear_object` to set the pointers to `NULL`) and disconnect signal handlers, but there's no point in freeing strings here. Do it in finalize instead; then you can use `g_free` directly instead of `g_clear_pointer`, since the object cannot be used again after it's finalized, and you don't have to worry about setting the pointer to `NULL`. (Remember that dispose can be called multiple times in a row, and the object must be left in a valid state after each call, so your all your code has to be prepared to handle a `NULL` `self->user` everywhere after construction if you free it here; you don't have to worry about that if you do it in finalize.)